### PR TITLE
Update API

### PR DESCRIPTION
**Pull request by ruthenian8**
# Description

Implement asynchronous execution of actor handlers, processing and response functions

# Checklist

- [x] I have covered the code with tests
- [x] I have added comments to my code to help others understand it
- [x] I have updated the documentation to reflect the changes
- [x] I have performed a self-review of the changes

### CODE REVIEW

**Comment by RLKRo** [comment]
Since this PR changes processing function API, maybe we should also remove/deprecate `overwrite_current_node_in_processing` here?

**Comment by RLKRo** [comment]
Maybe make all user-defined functions async?
https://github.com/deeppavlov/dialog_flow_framework/discussions/237#discussioncomment-7228940 (the list of functions)

Also, maybe rename `wrap_sync_function_in_async` to something like `handle_user_function` and let it also accept a function type (e.g. `"response"`, or `"stats_extractor"`) and then that function would handle exceptions / incorrect return type.

**Review by RLKRo — CHANGES_REQUESTED** [review]

File: dff/pipeline/service/service.py
[the remark this answers is not in the document]
Comment (ruthenian8) [reply dff/pipeline/service/service.py:1]: Updated

File: dff/pipeline/pipeline/actor.py [hunk 1]
```diff
@@ -291,10 +369,17 @@ def _get_true_label(
...
+            )
+        )
+        for label, cond_is_true in zip(transitions, cond_booleans):
+            if cond_is_true:
                 if callable(label):
-                    label = label(ctx, pipeline, *args, **kwargs)
+                    label = await wrap_sync_function_in_async(label, ctx, pipeline, *args, **kwargs)
                     # TODO: explicit handling of errors
                     if label is None:
                         continue
```
[the remark this answers is not in the document]
Comment (ruthenian8) [reply dff/pipeline/pipeline/actor.py:387]: Updated

File: dff/pipeline/pipeline/actor.py [hunk 2]
```diff
@@ -109,51 +106,51 @@ def __init__(
...
 
         # rewrite previous node
         ctx = self._rewrite_previous_node(ctx, *args, **kwargs)
-        self._run_handlers(ctx, pipeline, ActorStage.REWRITE_PREVIOUS_NODE, *args, **kwargs)
+        await self._run_handlers(ctx, pipeline, ActorStage.REWRITE_PREVIOUS_NODE, *args, **kwargs)
 
         # run pre transitions processing
-        ctx = self._run_pre_transitions_processing(ctx, pipeline, *args, **kwargs)
-        self._run_handlers(ctx, pipeline, ActorStage.RUN_PRE_TRANSITIONS_PROCESSING, *args, **kwargs)
+        await self._run_pre_transitions_processing(ctx, pipeline, *args, **kwargs)
```
[the remark this answers is not in the document]
Comment (ruthenian8) [reply dff/pipeline/pipeline/actor.py]: Updated

[the remark this answers is not in the document]
Comment (ruthenian8) [reply dff/pipeline/pipeline/actor.py]: Updated

File: dff/pipeline/pipeline/actor.py [hunk 3]
```diff
@@ -262,25 +264,101 @@ def _overwrite_node(
...
+        **kwargs,
+    ) -> Context:
+        """
+        Executes the normalized response as an asynchronous function.
+        See the details in the :py:func:`~normalize_response` function of `normalization.py`.
+        """
+        response = normalize_response(response)
+        return await wrap_sync_function_in_async(response, ctx, pipeline, *args, **kwargs)
+
+    async def _run_processing_parallel(
```
[the remark this answers is not in the document]
Comment (ruthenian8) [reply dff/pipeline/pipeline/actor.py:287]: Updated

File: dff/pipeline/service/service.py [hunk 4]
```diff
@@ -7,12 +7,9 @@
 Service group can be synchronous or asynchronous.
 Service is an atomic part of a pipeline.
 Service can be asynchronous only if its handler is a coroutine.
-Actor wrapping service can be synchronous only.
+Actor wrapping service is asynchronous.
```
[the remark this answers is not in the document]
Comment (ruthenian8) [reply dff/pipeline/service/service.py:10]: is

File: dff/pipeline/pipeline/actor.py [hunk 5]
```diff
@@ -109,51 +111,51 @@ def __init__(
...
+    async def __call__(
         self, pipeline: Pipeline, ctx: Optional[Union[Context, dict, str]] = None, *args, **kwargs
     ) -> Union[Context, dict, str]:
         # context init
         ctx = self._context_init(ctx, *args, **kwargs)
-        self._run_handlers(ctx, pipeline, ActorStage.CONTEXT_INIT, *args, **kwargs)
+        await self._run_handlers(ctx, pipeline, ActorStage.CONTEXT_INIT, *args, **kwargs)
 
         # get previous node
         ctx = self._get_previous_node(ctx, *args, **kwargs)
```
[the remark this answers is not in the document]
Comment (ruthenian8) [reply dff/pipeline/pipeline/actor.py:122]: We can leave such methods as `get_previous_node` and `get_next_node` synchronous for two reasons:
1) all they do is get values from dictionaries, so making them asynchronous won't result in a performance improvement
2) we would have to await them inside the actor call, i.e. use them like regular blocking functions

**Review by pseusys — APPROVED** [review]
Looks good to me!

File: dff/pipeline/pipeline/pipeline.py [hunk 6]
```diff
@@ -13,16 +13,17 @@
...
-from typing import Union, List, Dict, Optional, Hashable, Callable
+from typing import Union, List, Dict, Optional, Hashable, Callable, TYPE_CHECKING
 
 from dff.context_storages import DBContextStorage
 from dff.script import Script, Context, ActorStage
 from dff.script import NodeLabel2Type, Message
 from dff.utils.turn_caching import cache_clear
 
-from dff.messengers.common import MessengerInterface, CLIMessengerInterface
+import dff.messengers.common
```
Comment (pseusys) [inline dff/pipeline/pipeline/pipeline.py:26]: Why can't we import the class itself (as we always do)?

File: docs/source/user_guides/context_guide.rst [hunk 7]
```diff
@@ -35,7 +35,7 @@ Let's consider some of the built-in callback instances to see how the context ca
       pattern = re.compile("[a-zA-Z]+")
 
       def regexp_condition_handler(
-          ctx: Context, pipeline: Pipeline, *args, **kwargs
+          ctx: Context, pipeline: Pipeline
       ) -> bool:
```
Comment (pseusys) [inline docs/source/user_guides/context_guide.rst:39]: Collapse to single line, maybe?

File: tutorials/script/core/9_pre_transitions_processing.py [hunk 8]
```diff
@@ -35,14 +35,14 @@
 
 # %%
 def save_previous_node_response_to_ctx_processing(
-    ctx: Context, _: Pipeline, *args, **kwargs
+    ctx: Context, _: Pipeline
 ) -> Context:
```
Comment (pseusys) [inline tutorials/script/core/9_pre_transitions_processing.py:39]: Collapse to one line?

**Review by pseusys — APPROVED** [review]

File: dff/pipeline/pipeline/pipeline.py (hunk 6 above)
Comment (RLKRo) [reply dff/pipeline/pipeline/pipeline.py]: For some reason I thought that there is a cyclic import between `pipeline.types` and `messengers.common.interface`. That is not the case.
Fixed in ce3a8a2f7936d0482cec7d14fa460ce3971e0ca3.

File: docs/source/user_guides/context_guide.rst (hunk 7 above)
Comment (RLKRo) [reply docs/source/user_guides/context_guide.rst]: Done in 48b69340ef8aaf0c7a229a47c7ee352d61f706cc.

File: tutorials/script/core/9_pre_transitions_processing.py (hunk 8 above)
Comment (RLKRo) [reply tutorials/script/core/9_pre_transitions_processing.py]: No, that would make lines longer than 80 characters, which would result in them not fitting into code blocks in the docs (#241).
An alternative is to rename these functions to have shorter names (e.g. `save_previous_node_response` and `prepend_previous_node_response`). This is done in d8075f0cab22bb43d5863dc17abd6519e3be97b9.

**Review by RLKRo — APPROVED** [review]

**Review by pseusys — APPROVED** [review]

**Pull request merged**